### PR TITLE
[SYCLomatic] Support migration for cusparse<T>csrgemm

### PR DESCRIPTION
**Pull request by zhiweij1**
Signed-off-by: Jiang, Zhiwei <zhiwei.jiang@intel.com>

### CODE REVIEW

File: clang/runtime/dpct-rt/include/sparse_utils.hpp [hunk 1]
```diff
@@ -1165,6 +1196,259 @@ inline void spsv(sycl::queue queue, oneapi::mkl::transpose trans_a,
...
+  oneapi::mkl::sparse::matrix_handle_t matrix_handle_b = nullptr;
+  oneapi::mkl::sparse::matrix_handle_t matrix_handle_c = nullptr;
+  oneapi::mkl::sparse::init_matrix_handle(&matrix_handle_a);
+  oneapi::mkl::sparse::init_matrix_handle(&matrix_handle_b);
+  oneapi::mkl::sparse::init_matrix_handle(&matrix_handle_c);
+
+  float *val_a = (float *)dpct::dpct_malloc(nnz_a * sizeof(float));
+  float *val_b = (float *)dpct::dpct_malloc(nnz_b * sizeof(float));
+  dpct::dpct_memset(val_a, 1, nnz_a * sizeof(float));
+  dpct::dpct_memset(val_b, 1, nnz_b * sizeof(float));
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/sparse_utils.hpp]: let's fill val_a/b with int(1).

File: clang/runtime/dpct-rt/include/sparse_utils.hpp [hunk 2]
```diff
@@ -1165,6 +1196,259 @@ inline void spsv(sycl::queue queue, oneapi::mkl::transpose trans_a,
...
+  oneapi::mkl::sparse::matrix_handle_t matrix_handle_a = nullptr;
+  oneapi::mkl::sparse::matrix_handle_t matrix_handle_b = nullptr;
+  oneapi::mkl::sparse::matrix_handle_t matrix_handle_c = nullptr;
+  oneapi::mkl::sparse::init_matrix_handle(&matrix_handle_a);
+  oneapi::mkl::sparse::init_matrix_handle(&matrix_handle_b);
+  oneapi::mkl::sparse::init_matrix_handle(&matrix_handle_c);
+
+  float *val_a = (float *)dpct::dpct_malloc(nnz_a * sizeof(float));
+  float *val_b = (float *)dpct::dpct_malloc(nnz_b * sizeof(float));
+  dpct::dpct_memset(val_a, 1, nnz_a * sizeof(float));
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/sparse_utils.hpp]: Please consult to MKL team on perf.

Comment (zhiweij1) [reply clang/runtime/dpct-rt/include/sparse_utils.hpp]: OK

File: clang/runtime/dpct-rt/include/sparse_utils.hpp (hunk 1 above)
Comment (zhiweij1) [reply clang/runtime/dpct-rt/include/sparse_utils.hpp]: done

**Review by spencerpatty — APPROVED** [review]
these changes generally look fine to me.

File: clang/runtime/dpct-rt/include/sparse_utils.hpp [hunk 3]
```diff
@@ -1164,7 +1195,228 @@ inline void spsv(sycl::queue queue, oneapi::mkl::transpose trans_a,
...
+/// \param [in] info_b Matrix info of the matrix B.
+/// \param [in] nnz_b Non-zero elements number of matrix B.
+/// \param [in] row_ptr_b An array of length row number + 1.
+/// \param [in] col_ind_b An array containing the column indices in index-based
+/// numbering.
+/// \param [in] info_c Matrix info of the matrix C.
+/// \param [in] row_ptr_c An array of length row number + 1.
+/// \param [out] nnz_host_ptr Non-zero elements number of matrix C.
+inline void
+csrgemm_nnz_estimiate(sycl::queue queue, oneapi::mkl::transpose trans_a,
```
Comment (spencerpatty) [inline clang/runtime/dpct-rt/include/sparse_utils.hpp]: check spelling here on "estimiate" -> "estimate"

File: clang/runtime/dpct-rt/include/sparse_utils.hpp [hunk 4]
```diff
@@ -1164,7 +1195,228 @@ inline void spsv(sycl::queue queue, oneapi::mkl::transpose trans_a,
...
+                              matrix_handle_c,
+                              oneapi::mkl::sparse::matmat_request::compute,
+                              matmat_descr, nullptr, nullptr
+#ifndef DPCT_USM_LEVEL_NONE
+                              ,
+                              {}
 #endif
+  );
+
+  if (is_estimation) {
```
Comment (spencerpatty) [inline clang/runtime/dpct-rt/include/sparse_utils.hpp]: so first time through would call 

is_estimation:  work_estimation + compute + get_nnz

second time through would call 

work_estimation + compute + finalize ??

Isn't this a little redundant and doing way more work than is necessary?  is csrgemm_nnz_estimate optional ?  or where does C matrix actually get allocated ? 

Can you actaully walk me through a use case of how it gets mapped ?

Comment (spencerpatty) [reply clang/runtime/dpct-rt/include/sparse_utils.hpp]: I would like to talk through this before you merge it.  I came back after the initial review and approval and found some questions still in my mind that I hadn't noticed before I approved.  :)

File: clang/runtime/dpct-rt/include/sparse_utils.hpp (hunk 3 above)
Comment (zhiweij1) [reply clang/runtime/dpct-rt/include/sparse_utils.hpp]: OK

File: clang/runtime/dpct-rt/include/sparse_utils.hpp (hunk 4 above)
Comment (zhiweij1) [reply clang/runtime/dpct-rt/include/sparse_utils.hpp]: Hi @spencerpatty , the example is:
image

Comment (zhiweij1) [reply clang/runtime/dpct-rt/include/sparse_utils.hpp]: And the full example files are:
[example.zip](https://github.com/oneapi-src/SYCLomatic/files/13507524/example.zip)

**Pull request closed without merging**